wpengine/faustjs
# FaustNX CLI POC

The `faustnx` CLI will allow us to implement telemetry, and better position us in the future for other features that require our input before the `next` CLI runs.

## Description

This PR introduces a `faustnx` CLI that acts as a pass through to the `next` CLI, however, with our own telemetry that runs prior to the `next` CLI.

When running the `faustnx` CLI for the first time, you are asked if you would like to enroll in FaustNX Telemetry. By responding with either `true` or `false`, your preferences are then saved to your local machine (`~/.config/configstore/faustnx.json` on Mac) as a JSON file:

```jsonc
{
  telemetry: {
    notifiedAt: '12345', // timestamp the user was first notified about FaustNX telemetry
    anonymousId: '0000-0000-0000', // An anonymous v4 UUID for GA
    enabled: true // Either true or false based on user preferences
  }
}
```

From there, any time the `faustnx` CLI is ran, we check the users' preferences, and if telemetry is enabled, we send a GA event with relevant information.

In addition to the basic Next CLI commands like `faustnx start`, `faustnx dev`, and `faustnx build`, there is also a new command `faustnx faustnx-telemetry`, which allows the user to modify their telemetry preferences.

## Testing

1. Checkout `faustnx-cli` branch
2. Run `npm run build -w faust-nx` from the monorepo root
3. Run `npm run build -w faust-nx-cli` from the monorepo root
4. Run `npm run dev -w faust-nx-getting-started` to start the FaustNX example app in dev mode
5. Notice the FaustNX Telemetry prompt
6. Type `true`
7. Exit out of `dev` mode
8. Run `npm run build -w faust-nx-getting-started`
9. Based on your preference, you will see a telemetry event being `console.log`'d
10. Now, run `npx faustnx faustnx-telemetry` and change your preferences to `false`
11. Run `npm run build -w faust-nx-getting-started` notice that the telemetry event is not being `console.log`'d

**NOTE**: WordPress telemetry info will not be collected and will display as `undefined` since the WP telemetry endpoint doesn't exist on this branch.

## Telemetry Data

Currently the following data is being captured in a telemetry event:

```ts
export interface TelemetryData {
  wordpress: {
    faustwp_settings: {
      has_frontend_uri?: boolean;
      redirects_enabled?: boolean;
      rewrites_enabled?: boolean;
      themes_disabled?: boolean;
      image_source_replacement_enabled?: boolean;
    };
    faustwp_version?: string;
    is_wpe?: boolean;
    multisite?: boolean;
    php_version?: string;
    wp_version?: string;
  };
  node: {
    faustnx_version?: string;
    faustnx_cli_version?: string;
    apollo_client_version?: string;
    node_version?: string;
    next_version?: string;
    is_development?: boolean;
  };
}
```
